crates/ide-assists/src/utils.rs

```diff
@@ -316,43 +316,52 @@ pub(crate) fn does_pat_match_variant(pat: &ast::Pat, var: &ast::Pat) -> bool {
     pat_head == var_head
 }
 
-pub(crate) fn does_nested_pattern(pat: &ast::Pat) -> bool {
-    let depth = calc_depth(pat, 0);
+pub(crate) fn does_pat_variant_nested_or_literal(pat: &ast::Pat) -> bool {
+    check_pat_variant_nested_or_literal_with_depth(pat, 0)
+}
 
-    if 1 < depth {
+fn check_pat_variant_nested_or_literal_with_depth(pat: &ast::Pat, depth: usize) -> bool {
+    if depth > 1 {
         return true;
     }
-    false
-}
 
-fn calc_depth(pat: &ast::Pat, depth: usize) -> usize {
     match pat {
         ast::Pat::IdentPat(_)
-        | ast::Pat::BoxPat(_)
         | ast::Pat::RestPat(_)
-        | ast::Pat::LiteralPat(_)
-        | ast::Pat::MacroPat(_)
-        | ast::Pat::OrPat(_)
-        | ast::Pat::ParenPat(_)
-        | ast::Pat::PathPat(_)
         | ast::Pat::WildcardPat(_)
-        | ast::Pat::RangePat(_)
-        | ast::Pat::RecordPat(_)
         | ast::Pat::RefPat(_)
-        | ast::Pat::SlicePat(_)
-        | ast::Pat::TuplePat(_)
-        | ast::Pat::ConstBlockPat(_) => depth,
+        | ast::Pat::BoxPat(_)
+        | ast::Pat::ConstBlockPat(_) => false,
+
+        ast::Pat::LiteralPat(_)
+        | ast::Pat::RangePat(_)
+        | ast::Pat::MacroPat(_)
+        | ast::Pat::PathPat(_) => true,
+
+        ast::Pat::ParenPat(pat) => pat
+            .pat()
+            .map_or(true, |pat| check_pat_variant_nested_or_literal_with_depth(&pat, depth)),
+        ast::Pat::TuplePat(pat) => {
+            pat.fields().any(|pat| check_pat_variant_nested_or_literal_with_depth(&pat, depth))
+        }
+        ast::Pat::RecordPat(pat) => pat.record_pat_field_list().map_or(true, |pat| {
+            pat.fields().any(|pat| {
+                pat.pat()
+                    .map_or(true, |pat| check_pat_variant_nested_or_literal_with_depth(&pat, depth))
+            })
+        }),
 
-        // FIXME: Other patterns may also be nested. Currently it simply supports only `TupleStructPat`
         ast::Pat::TupleStructPat(pat) => {
-            let mut max_depth = depth;
-            for p in pat.fields() {
-                let d = calc_depth(&p, depth + 1);
-                if d > max_depth {
-                    max_depth = d
-                }
-            }
-            max_depth
+            pat.fields().any(|pat| check_pat_variant_nested_or_literal_with_depth(&pat, depth + 1))
+        }
+        ast::Pat::SlicePat(pat) => {
+            pat.pats().next().is_none()
+                || pat
+                    .pats()
+                    .any(|pat| check_pat_variant_nested_or_literal_with_depth(&pat, depth + 1))
```

Comment:
It is not clear to me what you are trying to check, but slices patterns are not easy to check. The only pattern that should return `false` for slices is `[..]`, but arrays can have more, and this is complicated to check. I think this should just return `true`. Anyway I was unable to demonstrate wrong behavior because the depth is always >1 (`Some([a])`), so maybe this is less important, but still.